src/linear_algebra/bilinear_form.lean

```diff
@@ -0,0 +1,226 @@
+(bilin_smul_right : ∀ (a : α) (x y : β), bilin x (a • y) = a * (bilin x y))
+
+namespace bilin_form
+ 
+open bilin_form
+
+variables {α : Type u} {β : Type v} [ring α] [add_comm_group β] [module α β] {B : bilin_form α β}
+
+instance module : module α β := by apply_instance
+
+lemma zero_bilin (x : β) :
+B.bilin 0 x = 0 := by rw [←@zero_smul α _ _ _ _ (0 : β), bilin_smul_left, zero_mul]
+
+lemma bilin_zero (x : β) :
+B.bilin x 0 = 0 := by rw [←@zero_smul _ _ _ _ _ (0 : β), bilin_smul_right, ring.zero_mul]
+
+lemma bilin_neg_left (x y : β) :
+B.bilin (-x) y = -(B.bilin x y) := by rw [←@neg_one_smul α _ _, bilin_smul_left, neg_one_mul]
+
+lemma bilin_neg_right (x y : β) :
+B.bilin x (-y) = -(B.bilin x y) := by rw [←@neg_one_smul α _ _, bilin_smul_right, neg_one_mul]
+
+lemma bilin_sub_left (x y z : β) :
+B.bilin (x - y) z = B.bilin x z - B.bilin y z := by rw [sub_eq_add_neg, bilin_add_left, bilin_neg_left]; refl
+
+lemma bilin_sub_right (x y z : β) :
+B.bilin x (y - z) = B.bilin x y - B.bilin x z := by rw [sub_eq_add_neg, bilin_add_right, bilin_neg_right]; refl 
+
+variable {D : bilin_form α β}  
+@[extensionality] lemma ext (H : ∀ (x y : β), B.1 x y = D.1 x y) : B = D := begin cases B, cases D, simp, funext, exact H _ _ end 
+
+instance : add_comm_group (bilin_form α β) :=
+{ add := λ B D, { bilin := λ x y, B.1 x y + D.1 x y,
+                  bilin_add_left := λ x y z, by {rw bilin_add_left, rw bilin_add_left, simp},
+                  bilin_smul_left := λ a x y, by {rw [bilin_smul_left, bilin_smul_left, mul_add]}, 
+                  bilin_add_right := λ x y z, by {rw bilin_add_right, rw bilin_add_right, simp},
+                  bilin_smul_right := λ a x y, by {rw [bilin_smul_right, bilin_smul_right, mul_add]}
+                  }, 
+  add_assoc := by {intros, ext, unfold bilin, rw add_assoc},
+  zero := { bilin := λ x y, 0,
+            bilin_add_left := λ x y z, eq_comm.mp (add_zero 0),
+            bilin_smul_left := λ a x y, eq_comm.mp (mul_zero a),
+            bilin_add_right := λ x y z, eq_comm.mp (zero_add 0),
+            bilin_smul_right := λ a x y, eq_comm.mp (mul_zero a),
+            },
+  zero_add := by {intros, ext, unfold bilin, rw zero_add},
+  add_zero := by {intros, ext, unfold bilin, rw add_zero},
+  neg := λ B, { bilin := λ x y, - (B.1 x y), 
+                bilin_add_left := λ x y z, by rw [bilin_add_left, neg_add],
+                bilin_smul_left := λ a x y, by rw [bilin_smul_left, mul_neg_eq_neg_mul_symm],
+                bilin_add_right := λ x y z, by rw [bilin_add_right, neg_add],
+                bilin_smul_right := λ a x y, by rw [bilin_smul_right, mul_neg_eq_neg_mul_symm],
+                },
+  add_left_neg := by {intros, ext, unfold bilin, rw neg_add_self},
+  add_comm := by {intros, ext, unfold bilin, rw add_comm},
+}
+
+section 
+
+variables {χ : Type*} [comm_ring χ] [module χ β] (F : bilin_form χ β) (f : β → β)
+
+instance base_ring_module : module χ χ := ring.to_module 
```

Comment:
This is already in mathlib, as you can see from your proofs. What is the reason for these instances?

Reply:
Without them the type class inference seems to get lost

Reply:
If you can prove this instance with `by apply_instance` then prove it like that, or else you're in danger of having more than one instance clogging up the system